### ISSUE TITLE
Input lag

### ISSUE THREAD

**Issue by tonymonsalo** [ValveSoftware/Dota2-Gameplay#15559]
### Ability name

basic dota behaviour

### Description

I hear about this issue around last 10 days from streamers, youtubers, on forums and experience this myself. But i don't see any reports on this bug-tracker.
There is significantly annoying input lag, mostly on movement and rotating. At the beginning I had it only on juggernaut, but now i notice it on more heroes (applied video with templar assassin). 

Some streamers said, that it is related to video settings, but it didn't helped me (maybe only a bit). As for me - i use borderless window (for the last7 years). Setting Fullscreen, i guess helps a bit, but not completely (and it is not comfortable)

Applying video, where i tried to synchronize hero movement in test lobby (left) and in mm game (right; video captured during a game).

https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/6010beda-7f23-4198-9ff5-bea65b539cc8

![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/106998ef-81b3-4903-9d61-d5e8b4d63b2a)
![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/a8150f4a-f3eb-48e8-af40-6c0b4978ff6a)

### Example Match ID (and possibly Timestamp)

7597571284 - 0:27 (this report is not about some separate game)

**Comment by kvaddballe** [comment]
This wasn't fixed in 7.35c unfortunately, I just played a game and I'm experiencing the exact same issue.

Adding threads from reddit that list this same issue:
https://www.reddit.com/r/dota2/comments/1awdid1/game_runs_horrible_since_chinese_new_year_patch/

https://www.reddit.com/r/dota2/comments/1au645l/please_fix_the_input_lag/

https://www.reddit.com/r/dota2/comments/1at9cmj/anyone_experiencing_input_lags/

https://www.reddit.com/r/dota2/comments/1atrfde/psa_use_exclusive_fullscreen_as_display_mode/

https://www.reddit.com/r/dota2/comments/1at7dlf/input_lag/

https://www.reddit.com/r/dota2/comments/1aszhl2/dota_2_delay/

https://www.reddit.com/r/dota2/comments/1asb3mc/input_lag_since_last_update/

https://www.reddit.com/r/dota2/comments/1arswv7/input_lag/

https://www.reddit.com/r/dota2/comments/1ap0sg8/some_microlags_on_servers/

https://www.reddit.com/r/dota2/comments/1anr5m5/mouse_input_lagdelay/

https://www.reddit.com/r/dota2/comments/1au29sc/very_slight_input_lag_after_10_mins_ingame/

**Comment by tonymonsalo** [comment]
@kvaddballe it's hard to believe, that players don't complain about this thing. And it's hard to believe that people accepted this as it's okay

**Comment by K-o-patra** [comment]
Yes I have noticed this as well. This happened the first time for me a couple of days ago, might have coincided with the dragon gift introduction. And it has been the same ballpark of lag, 200-300 ms at most times. Mostly at the start of the game, within the first few minutes.

**Comment by quinnbller5** [comment]
Same thing for me, i play on US east and have ~10 ping, it feels like its 80-100. Certain games it feels worse than others, ive tried every settings shuffle to fix it, nothing seems to work consistently, not sure what to do

**Comment by Razer725** [comment]
Not fixed after patch 7.35c. The heroes move as if under sticky napalm.

**Comment by tonymonsalo** [comment]
> Not fixed after patch 7.35c. The heroes move as if under sticky napalm.

well said

**Comment by jeffhill** [comment]
For folks who are seeing this, could you try "cl_clock_recvmargin_enable 0" in the console? It *might* fix this problem, and if so it narrows down the bug quite a lot. Thank you!

**Comment by tonymonsalo** [comment]
> For folks who are seeing this, could you try "cl_clock_recvmargin_enable 0" in the console? It _might_ fix this problem, and if so it narrows down the bug quite a lot. Thank you!

no, this didn't helped

**Comment by jeffhill** [comment]
Could you share a MatchID and which hero you played when you saw the problem? I'd like to pull logs from the servers to check a few things. Thank you!

**Comment by tonymonsalo** [comment]
> Could you share a MatchID and which hero you played when you saw the problem (very recent MatchIDs help)? I'd like to pull logs from the servers to check a few things. Thank you!

7601901465
Played just minutes ago

**Comment by tonymonsalo** [comment]
> > Could you share a MatchID and which hero you played when you saw the problem (very recent MatchIDs help)? I'd like to pull logs from the servers to check a few things. Thank you!
> 
> 7601901465 Played just minutes ago It was chaos knight

@jeffhill Also, I entered "cl_clock_recvmargin_enable 0" during that match. I did it at around 40 minute (and commented to you right after)

**Comment by Wandang** [comment]
MatchID: 7601966494
It seemed to have helped on my end. I was the alchemist

**Comment by Wandang** [comment]
> > For folks who are seeing this, could you try "cl_clock_recvmargin_enable 0" in the console? It _might_ fix this problem, and if so it narrows down the bug quite a lot. Thank you!
> 
> no, this didn't helped

are you sure it was activated?

for those that don't know: you need to add it with a + into the launch options for it to work:

```
+cl_clock_recvmargin_enable 0
``

**Comment by tonymonsalo** [comment]
> are you sure it was activated?
> 
> for those that don't know: you need to add it with a + into the launch options for it to work:
> 
> ```
> +cl_clock_recvmargin_enable 0
> ```

Ah ok, I was entering this in game console. 
Just in case, this is console log when i did it:
```
> cl_clock_recvmargin_enable 0
[Client] Server tick 75982 was long (36.4ms)
[SndOperators] opvar_get_float operator import_random_entry_start::get_no_repeat_opvar in dota_battleend failed to get field dota_change_this_opvar_name
[SndOperators] opvar_get_float operator import_random_entry_start::get_no_repeat_opvar in dota_battleend failed to get field dota_change_this_opvar_name
[Client] Server tick 76682 was long (36.1ms)
```

Now i will restart game and play a game with this launch option
![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/3c2104b5-d7eb-43f6-87ff-56b83ced40ba)

**Comment by kvaddballe** [comment]
Here are two IDs I just played where it lags:
7601872537 (I'm Vengeful Spirit)
7601954295 (I'm Windranger)

**Comment by kvaddballe** [comment]
Launching the game now with this to see if it works: 
![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/160855876/f925ed25-2ac2-4097-b4f8-62b285dcb563)

Previously I just entered "cl_clock_recvmargin_enable 0" in the console directly in game, which didn't help (had it typed in for the two games above).

**Comment by Wandang** [comment]
> Launching the game now with this to see if it works: ![image](https://private-user-images.githubusercontent.com/160855876/307141565-f925ed25-2ac2-4097-b4f8-62b285dcb563.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2MzQ3MzUsIm5iZiI6MTcwODYzNDQzNSwicGF0aCI6Ii8xNjA4NTU4NzYvMzA3MTQxNTY1LWY5MjVlZDI1LTJhYzItNDA5Ny1iNGY4LTYyYjI4NWRjYjU2My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyMlQyMDQwMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wODNiMGNlZjY5ZDYyZDNjYWM4ZjBlOTI5MmQ5ZmQyNThjZmNiMDViZmZiZmJhODlhOGM3MDNkMTliOTUyZDE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4JdYLpF-0wITOn6IEaJnMf3KusaU9xTeA93ElCVxDSk)
> 
> Previously I just entered "cl_clock_recvmargin_enable 0" in the console directly in game, which didn't help (had it typed in for the two games above).
> 
> Also: when I type in "cl_clock_recvmargin_enable" in console it comes back with "cl_clock_recvmargin_enable = false" - is this how it should be Jeff?

yes, in steam console 0 = false, 1 = true

**Comment by jeffhill** [comment]
Yeah, sounds like cl_clock_recvmargin_enable isn't related to the problem. Boo. Thanks for testing it!

**Comment by tonymonsalo** [comment]
@jeffhill @Wandang 
Thank you guys. Can i close this Issue?

**Comment by kvaddballe** [comment]
Hmm. Tried a game now with it entered before I launched the game. I think it's improved, but doesn't feel fully responsive still. My FPS also seems lower than before, hovering around 100. What exactly does the setting do @jeffhill ?

**Comment by jeffhill** [comment]
There was a change to some clock sync / frame pacing code that shipped with the last update. Setting this value to 0 makes it use the old code path from before the update.

@tonymonsalo @kvaddballe  can you confirm that running Dota with "+cl_clock_recvmargin_enable 0" on the command line solves the problem for you?

We shouldn't close this until it's fixed for real, just so there's a place to track the testing efforts.

**Comment by zenkovsan** [comment]
Just played a game using +cl_clock_recvmargin_enable 0 launch option. Input was much better than before.

**Comment by zenkovsan** [comment]
But still, I'm not sure if it was placebo effect or real change.

**Comment by kvaddballe** [comment]
Just played a second game with the setting and I can confirm it feels better than before, but I'm also very certain that there's still more input lag than previously before patch and the game still doesn't feel good. At times I had similar levels of input lag, but overall in the game it was less. Still doesn't feel "crisp" as it's always done previously.

Game ID: 7602115618 (I was storm)

**Comment by gmackall** [comment]
Also just to be clear, if the "last patch" is referring to 7.35c, then I (and judging by the timing of the reddit posts, many others) experienced the lag before 7.35c (but if it is referring to a different unnamed patch, then carry on!)

**Comment by jeffhill** [comment]
>Just to be clear, if the "last update" is referring to 7.35c
The clock change related to cl_clock_recvmargin_enable shipped in the Feb 8th update, so I *think* the timelines line up.

**Comment by jeffhill** [comment]
Thanks, @kvaddballe . Pulling server and various backend logs, I can confirm that the game server saw a very stable 34ms ping connection to you for the whole match, with only 3 lost packets the whole game. That test rules out an awful lot of possible causes, thank you!

**Comment by K-o-patra** [comment]
I have two different internet connections in parallel. And this issue arises on both of them. I have two connections just to rule out ISPs and infrastructure as possible explanations for the issue called [micro lags](https://github.com/ValveSoftware/Dota2-Gameplay/issues/15078), but this one also happens on both of my internet connections. These are totally different connections, from different ISPs, using two different networking technologies, and totally different physical signal transmitter mediums, as described [here](https://github.com/ValveSoftware/Dota2-Gameplay/issues/15078#issuecomment-1944100299). Does this extra weight to the findings from the previous comment?

**Comment by tonymonsalo** [comment]
> There was a change to some clock sync / frame pacing code that shipped with the last update. Setting this value to 0 makes it use the old code path from before the update.
> 
> @tonymonsalo @kvaddballe can you confirm that running Dota with "+cl_clock_recvmargin_enable 0" on the command line solves the problem for you?
> 
> We shouldn't close this until it's fixed for real, just so there's a place to track the testing efforts.

Overall "+cl_clock_recvmargin_enable 0" makes things better. But i still feel that something is off 
(i play 5-10 matches everyday for the last 7 years, so i think i feel it 😵‍💫 ; but maybe my eye is blurred already)

**Comment by kvaddballe** [comment]
> > There was a change to some clock sync / frame pacing code that shipped with the last update. Setting this value to 0 makes it use the old code path from before the update.
> > @tonymonsalo @kvaddballe can you confirm that running Dota with "+cl_clock_recvmargin_enable 0" on the command line solves the problem for you?
> > We shouldn't close this until it's fixed for real, just so there's a place to track the testing efforts.
> 
> Overall "+cl_clock_recvmargin_enable 0" makes things better. But i still feel that something is off (i play 5-10 matches everyday for the last 7 years, so i think i feel it 😵‍💫 ; but maybe my eye is blurred already)
> 
> Changing hero move direction feels delayed

Ok thank god, it's not just in my head either then. I've kept playing games tonight and can definitely conclude that the issue persists even with "+cl_clock_recvmargin_enable 0", but it alleviates it somewhat. It's clear as day when playing certain heroes such as Timbersaw, Storm Spirit, Primal Beast, Tiny that turn around alot and have alot of inputs in short time. Also very noticeable if you try to fly around in circles with the flying courier. The difference becomes very stark when you compare Demo mode to MM.

Another game ID if it's of any use Jeff: 7602212457 (I'm Storm Spirit).

**Comment by kvaddballe** [comment]
> Thanks, @kvaddballe . Pulling server and various backend logs, I can confirm that the game server saw a very stable 34ms ping connection to you for the whole match, with only 3 lost packets the whole game. That test rules out an awful lot of possible causes, thank you!

Happy to be of help, let us know if there's anything else we can do to give you data that can be helpful in resolving the issue.

**Comment by zenkovsan** [comment]
Pangolier has 0 turn rate, so it is better to test using Pango

**Comment by IgorHoruzha** [comment]
> I have two different internet connections in parallel. And this issue arises on both of them. I have two connections just to rule out ISPs and infrastructure as possible explanations for the issue called [micro lags](https://github.com/ValveSoftware/Dota2-Gameplay/issues/15078), but this one also happens on both of my internet connections. These are totally different connections, from different ISPs, using two different networking technologies, and totally different physical signal transmitter mediums, as described [here](https://github.com/ValveSoftware/Dota2-Gameplay/issues/15078#issuecomment-1944100299). Does this rule out ISPs and infrastructure as possible explanations?

Have You tried using Your mobile phone(4G Ethernet) as a USB modem?

**Comment by kvaddballe** [comment]
Someone on reddit posted this as a possible solution: https://www.reddit.com/r/DotA2/comments/18ffzp6/script_that_may_fix_issues_you_are_having_with/

It's a script that resets all your settings as if your account is "new" again. I tried it but it didn't work unfortunately.

**Comment by Fivedota** [comment]
> Could you share a MatchID and which hero you played when you saw the problem (very recent MatchIDs help)? I'd like to pull logs from the servers to check a few things. Thank you!

I will give you two where I clearly remember that something was so off:
76002600100 (I'm the Juggernaut)
7594869912 (I'm the DP)

**Comment by SpiritWinD86** [comment]
Can someone re-test this command (disable\enable) on one server (so that the ping still the same, for example in lobby on the server) and **please, confirm if it actually works or not.**
Players have already started recommending this option\command on the forums\reddit and we just don't need another misinformation with the launch options or console commands that's not really works, just to make sure.

For those who's copying this command to the console, it will look like this: **cl_clock_recvmargin_enable false** (or 0, looks like it works in both ways)
or like this in the launch options of the game's properties in steam: **+cl_clock_recvmargin_enable 0** (or false). 

I tried this command and for me it looks like another placebo effect. Either it works or it doesn’t, and I don’t see much of a difference. 

After release of The Dragon's gift update in the first game, I had delay (input lag) with pressing buttons and hero's response like around one to two seconds, and after that ruined game i'm completely reinstalled dota (by deleting game's folder and userdata\570) and it became better, but it still feels like something is wrong. Controlling the hero's movement and abilities is not as responsive as before. Playing on EU servers with 60-85 ping (same as before).

**Comment by kvaddballe** [comment]
> Can someone re-test this command (disable\enable) on one server (so that the ping still the same, for example in lobby on the server) and **please, confirm if it actually works or not.** Players have already started recommending this option\command on the forums\reddit and we just don't need another misinformation with the launch options or console commands that's not really works, just to make sure.
> 
> For those who's copying this command to the console, it will look like this: **cl_clock_recvmargin_enable false** (or 0, looks like it works in both ways) or like this in the launch options of the game's properties in steam: **+cl_clock_recvmargin_enable 0** (or false).
> 
> I tried this command and for me it looks like another placebo effect. Either it works or it doesn’t, and I don’t see much of a difference.
> 
> After release of The Dragon's gift update in the first game, I had delay (input lag) with pressing buttons and hero's response like around one to two seconds, and after that ruined game i'm completely reinstalled dota (by deleting game's folder and userdata\570) and it became better, but it still feels like something is wrong. Controlling the hero's movement and abilities is not as responsive as before. Playing on EU servers with 60-85 ping (mostly on West EU, same ping as before).

I have the launch command and can confirm it does NOT resolve the issue. It may slightly improve it, but that might as well be placebo as you mentioned. I am however 100% sure that I'm still experiencing quite heavy input lag, noticably like before.

Playing on EU servers.

**Comment by khayuri** [comment]
haven't tested the jeff recommended command but one weird behaviour i have noticed is, if i go to speedtest.net and do an internet test, for some reason when im back into the game (while in a ranked game and the input lag starts showing up) the input lag is gone for like seconds/minutes but then it just comes back, with different intensities (atleast thats how i perceive it)
also applying video settings again even though i dont change anythign somehow still aleviates the problem a little bit.

**Comment by kvaddballe** [comment]
> haven't tested the jeff recommended command but one weird behaviour i have noticed is, if i go to speedtest.net and do an internet test, for some reason when im back into the game (while in a ranked game and the input lag starts showing up) the input lag is gone for like seconds/minutes but then it just comes back, with different intensities (atleast thats how i perceive it) also applying video settings again even though i dont change anythign somehow still aleviates the problem a little bit.

That does sound weird. However, it could just be the fact that you're alt-tabbing that is making it better for a few minutes? A lot of users (me included) have reported that alt-tabbing (sometimes a few times back and forth) removes the input lag for a few minutes.

**Comment by Fivedota** [comment]
Now I have heard Qojqva, Insania and W33 all talked about it on their stream and they said that a lot of players have this issue. Even rank 12 GPK on EU:s nickname rightnow is "Input Lag Enjoyer" so I mean, why can't we have a proper fix for this. No console commands, no fixes where I have to reset my whole steam account, no disable fullscreen opt or w/e. Just fix the issue already Valve please

**Comment by evilch3n** [comment]
just wanted to add to the discussion - affects me and my mate randomly too - its very frustrating

ive added +cl_clock_recvmargin_enable 0 to my launch options but it still feels off but not as extreme? (might be placebo tho)

noticable if u try to press more than 2 things really quick like a item + skill1 + skill2 in succession in hectic fights

**Comment by jeffhill** [comment]
For players who are seeing the problem, could you try running with "-condebug" on the command line, playing a bit on a server to have the problem, and then share the console.log file that's generated during the game?

We're currently running with "+cl_clock_recvmargin_enable 0" on by default in the launcher for everyone, but it's still unclear if that actually helps.

Thanks for your patience folks - it's tricky to track down how to reproduce problems like this that happen 100% of the time for some unlucky players, and not at all for many others.

**Comment by tonymonsalo** [comment]
> For players who are seeing the problem, could you try running with "-condebug" on the command line, playing a bit on a server to have the problem, and then share the console.log file that's generated during the game?
> 
> We're currently running with "+cl_clock_recvmargin_enable 0" on by default in the launcher for everyone, but it's still unclear if that actually helps.
> 
> Thanks for your patience folks - it's tricky to track down how to reproduce problems like this that happen 100% of the time for some unlucky players, and not at all for many others.

Started around 20:56

[console.log](https://github.com/ValveSoftware/Dota2-Gameplay/files/14424662/console.log)

**Comment by jeffhill** [comment]
@tonymonsalo Is that log file getting cut off by GitHub? It ends abruptly just as we're getting to the interesting part of the game... Any chance you could put it on pastebin or something? Thank you!

**Comment by tonymonsalo** [comment]
> @tonymonsalo Is that log file getting cut off by GitHub? It ends abruptly just as we're getting to the interesting part of the game... Any chance you could put it on pastebin or something? Thank you!

I posted it just during the game, this version should be full:

[console.log](https://github.com/ValveSoftware/Dota2-Gameplay/files/14425052/console.log)

**Comment by tonymonsalo** [comment]
> > @tonymonsalo Is that log file getting cut off by GitHub? It ends abruptly just as we're getting to the interesting part of the game... Any chance you could put it on pastebin or something? Thank you!
> 
> I posted it just during the game, this version should be full:
> 
> [console.log](https://github.com/ValveSoftware/Dota2-Gameplay/files/14425052/console.log)

@jeffhill 

Ahh... Sorry. Looks like it was cleared after restarting the game..

**Comment by tonymonsalo** [comment]
@jeffhill 
Here is .log to the end of the game
[console - Copy (3).log](https://github.com/ValveSoftware/Dota2-Gameplay/files/14425489/console.-.Copy.3.log)

@kvaddballe 
i passed -condebug in launch options

**Comment by kvaddballe** [comment]
> @jeffhill Here is .log to the end of the game [console - Copy (3).log](https://github.com/ValveSoftware/Dota2-Gameplay/files/14425489/console.-.Copy.3.log)
> 
> @kvaddballe i passed -condebug in launch options

And where does the log end up?

**Comment by tonymonsalo** [comment]
> > @jeffhill Here is .log to the end of the game [console - Copy (3).log](https://github.com/ValveSoftware/Dota2-Gameplay/files/14425489/console.-.Copy.3.log)
> 
> > 
> 
> > @kvaddballe i passed -condebug in launch options
> 
> 
> 
> And where does the log end up?

search "console.log" file in dota2 directory. I'm in bed already, so i can't give you exact path

**Comment by kvaddballe** [comment]
Doesn't seem to be working for me. Hmmm.

On another note. Someone on reddit posted that uninstalling and re-installing the Windows-updates from the 13/14th of February made the lag less bad. I've tried this and feel it's also abit lower. Can other people with input lag try this and see if it gets better for them?

Update names if ur on Windows 10: KB5034763 and KB5034685
Update names if ur in Windows 11: KB5034467 and KB5034765

**Comment by jeffhill** [comment]
Thank you for the console logs! I've done a bunch of digging into evidence in there, particularly these really confusing looking:
02/27 22:26:06 [Client] Server tick 35039 was not available when we needed it, arrived -65.3ms too late

I think what's happening here is that the client is hitting a spectacularly slow frame for some reason - longer than 33ms (30fps). It would really help if someone could install UI for ETW (https://github.com/google/UIforETW/releases) and capture an ETL file while the latency is happening? That should collect all the data needed to find the smoking gun on this one. We've already ruled out networking though the server logs showing stable, good connections and good server performance - so we know there's something weird happening inside the dota2.exe process that's causing this input latency and an ETL file would capture **everything** involved.

Thanks for all the help tracking this down folks!

**Comment by SpiritWinD86** [comment]
I think i've saw this "server tick" something.. and similar to this things in the console when game was laggy. It feels like your hero is constantly under sticky napalm, but it affects not only movement, but also pressing ability buttons.

And i don't think that it is windows related as mentioned above, because i'm playing on linux and faced same issue.

**Comment by jeffhill** [comment]
@tonymonsalo Could you try:
 - launch with "-condebug +cl_ticktiming 300 detail" on the command line
 - when you're in a match, in the in-game console try "cl_clock_recvmargin_enable 1". See if you encounter the lag?
 - If you hit the latency, try "cl_clock_recvmargin_enable 0" in the in-game console rather than alt-tabbing out.
 
Finally, if you could send me your console.log from this test, that would be super useful. I'll be able to see where in the pipeline the lag is coming from. Thank you!

**Comment by tonymonsalo** [comment]
> @tonymonsalo Could you try:
> 
> * launch with "-condebug +cl_ticktiming 300 detail" on the command line
> * when you're in a match, in the in-game console try "cl_clock_recvmargin_enable 1". See if you encounter the lag?
> * If you hit the latency, try "cl_clock_recvmargin_enable 0" in the in-game console rather than alt-tabbing out.
> 
> Finally, if you could send me your console.log from this test, that would be super useful. I'll be able to see where in the pipeline the lag is coming from. Thank you!

[console - Copy (4).log](https://github.com/ValveSoftware/Dota2-Gameplay/files/14453027/console.-.Copy.4.log) 

I was experiencing lag whether i change cl_clock_recvmargin_enable to 0 or 1 during the game (am i even allowed to do this multiple times?)

**Comment by tonymonsalo** [comment]
@jeffhill Also, i didn't get what you mean by "**rather than alt-tabbing out**"

**Comment by jeffhill** [comment]
Thanks for the log!

My understanding is that one "fix" folks are using is to alt+tab out of Dota and then back in to "reset" whatever's going wrong - and that would spoil the log data in this case.

Running a pre-update version of Dota on public would be pretty complicated as we'd need server infrastructure set up for that as well. The linked video at the top of this ticket shows 200ms latency on a 40ms connection, which is pretty noticeable.

**Comment by wazerstar** [comment]
> Thank you for the console logs! I've done a bunch of digging into evidence in there, particularly these really confusing looking: 02/27 22:26:06 [Client] Server tick 35039 was not available when we needed it, arrived -65.3ms too late
> 
> I think what's happening here is that the client is hitting a spectacularly slow frame for some reason - longer than 33ms (30fps). It would really help if someone could install UI for ETW (https://github.com/google/UIforETW/releases) and capture an ETL file while the latency is happening? That should collect all the data needed to find the smoking gun on this one. We've already ruled out networking though the server logs showing stable, good connections and good server performance - so we know there's something weird happening inside the dota2.exe process that's causing this input latency and an ETL file would capture **everything** involved.
> 
> Also helpful would be running the "cl_ticktiming" command in the console while you're having input lag and sharing the result. Comparing that to a "cl_ticktiming" command when you're _not_ having input lag in the early game would be very helpful too.
> 
> Thanks for all the help tracking this down folks!

You don't want to use Fletchers Dunn's guide?=

https://docs.google.com/document/d/1v8QtbUtGAf6OapE9iWn4Y4STA7Scn_hr1Yih1nf2_BA/edit#heading=h.u0cg8jbrukvb

**Comment by tonymonsalo** [comment]
> Thank you for the console logs! I've done a bunch of digging into evidence in there, particularly these really confusing looking: 02/27 22:26:06 [Client] Server tick 35039 was not available when we needed it, arrived -65.3ms too late
> 
> I think what's happening here is that the client is hitting a spectacularly slow frame for some reason - longer than 33ms (30fps). It would really help if someone could install UI for ETW (https://github.com/google/UIforETW/releases) and capture an ETL file while the latency is happening? That should collect all the data needed to find the smoking gun on this one. We've already ruled out networking though the server logs showing stable, good connections and good server performance - so we know there's something weird happening inside the dota2.exe process that's causing this input latency and an ETL file would capture **everything** involved.
> 
> Also helpful would be running the "cl_ticktiming" command in the console while you're having input lag and sharing the result. Comparing that to a "cl_ticktiming" command when you're _not_ having input lag in the early game would be very helpful too.
> 
> Thanks for all the help tracking this down folks!

@jeffhill  Didn't seen this comment

Here are some etl things
https://s3.eu-central-1.amazonaws.com/mists.life/2024-03-01_01-26-34_Konst.etl
https://s3.eu-central-1.amazonaws.com/mists.life/2024-03-01_01-56-31_Konst.etl
https://s3.eu-central-1.amazonaws.com/mists.life/2024-03-01_01-57-24_Konst.etl

Not sure that i am recorded them correctly, just clicked "Start recording" 
![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/f64c8658-35b6-4b96-ba3f-12d104554bae)

@wazerstar thanks for noticing. I guess, link that you provided leads for alternate way of recording ETL thing, and it is not necessary?

**Comment by jeffhill** [comment]
I'm still digging through those ETL files (thank you!) but it looks like your CPU is often idle for 10ms+ waiting on vsync. It looks like you're usually getting around 30 fps, which means that if even a single frame goes over the deadline and it has to wait on vsync it might be a full 16ms before that 60Hz vsync comes in - during which time everything else is idle.

**Comment by tonymonsalo** [comment]
> getting around 30 fps

Not really. I have vsync enabled, and it is mostly straight 59-60fps. Without vsync - it is 80-120fps.

(not sure that i understood meanings of all your words correctly 😅)

**Comment by jeffhill** [comment]
Oh, sorry, I misread the ETL. It's reporting 60fps, not 30fps.

Can you try adding "+dota_debug_change_cursor_during_pending_order 1" on the command line? This will make the mouse cursor change to the coaching cursor which is brighter, while you have an order that the server hasn't acknowledged yet. This will make it visually obvious when there's input lag, as your mouse will stay bright longer than normal. Basically every time you click the cursor will be bright until you hear back from the server that it got the order, so you'll be able to *see* any lag visually.

Another thing to try is setting "cl_interp_ratio 0". This effectively turns off the client side smoothing of data, which will make things more responsive, but at the cost of making any jitter due to lost packets more obvious.

**Comment by jeffhill** [comment]
We *just* shipped a client update that we believe fixes the client input latency bug. Please let me know if you see the problem in any new matches!

**Comment by Aeswaaaaa** [comment]
> We _just_ shipped a client update that we believe fixes the client input latency bug. Please let me know if you see the problem in any new matches!

Checking right now, testing to test with cl_clock_recvmargin_enable true or false? Or does it not matter

**Comment by Aeswaaaaa** [comment]
When I wake up I will test again with different values of cl_clock_recvmargin_enable, if I notice a lags pattern I will try to take a video clip through obs, if you need to measure indicators with certain software or inside the dota itself with commands, I can also do it, just let me know what's required of me.

**Comment by tonymonsalo** [comment]
> Oh, sorry, I misread the ETL. It's reporting 60fps, not 30fps.
> 
> Can you try adding "+dota_debug_change_cursor_during_pending_order 1" on the command line? This will make the mouse cursor change to the coaching cursor which is brighter, while you have an order that the server hasn't acknowledged yet. This will make it visually obvious when there's input lag, as your mouse will stay bright longer than normal. Basically every time you click the cursor will be bright until you hear back from the server that it got the order, so you'll be able to _see_ any lag visually.
> 
> Another thing to try is setting "cl_interp_ratio 0". This effectively turns off the client side smoothing of data, which will make things more responsive, but at the cost of making any jitter due to lost packets more obvious.

Played several games with "+dota_debug_change_cursor_during_pending_order 1". For me, looked like cursor always changed for a same amount of time (didn't seen any extra long duration of changed cursor).
As i understood, game unit should make their actions after the cursor changed to the initial "skin"?

I've tried to measure something

https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/4d770d7a-5b28-4a8f-9996-9c8897fdbafb

 (this video is also with cl_interp_ratio 0)
![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/d75018d2-ee8a-4084-9c71-007af33d375a)

**Comment by Aeswaaaaa** [comment]
> We _just_ shipped a client update that we believe fixes the client input latency bug. Please let me know if you see the problem in any new matches!

Do you have the ability to run old dota 2 builds locally on your computer? I just gathered some information, but to finalize it, I need to get an old build of Dota 2. Or that someone else has access to it.

**Comment by villagecapo** [comment]
Many people have unoptimized computers. I think it's not always about the game. There are many factors that affect input lag.

**Comment by mercdaddi** [comment]
> Many people have unoptimized computers. I think it's not always about the game. There are many factors that affect input lag. I tried "cl_clock_recvmargin_enable", almost the same. Nowadays, more emphasis is placed on input latency based on time of day.

my brotha whaat did you just type?

**Comment by Andreixzc** [comment]
Is your network settings set to 'high-end network'?
![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/90361670/5076d8ab-13a6-4ae6-a3c2-f0208a99cb64)

**Comment by mercdaddi** [comment]
Is everyone also getting the "excessive frame time clamped...acting as if the frame only took 100 MS?" I get that and the missing tick in the console in game. The frame time bug is worse, as it drops me from  230ish FPS to 15-50 for about 5 seconds. Happens much more when clicking on heroes and things are going on. It will happen 20-30 times a game, but then work like butter when not. It feels like it is alt-tabbing, but I dont have tab binded as tab. I made it LEFT arrow key so it wouldn't alt-tab out when I clicked it (I just use windows key and exclusive fullscreen)

I have tried every combo fathomable of fullscreen, borderless, and all their optimizations both on and off in windows.  I have unplugged/replugged every single monitor/device. I think I have tried all the console commands, but tbh I am not good in console and may have fucked them up, or maybe somehow made my game run worse. I have tried all NVIDIA settings in control panel, and combo of video settings in dota. I have tried system restore, and clean install of windows 11, Steam, and Dota. I prefer to use a 3 monitor setup. Not sure what else to do. I have not tried factory reset.

I will say, for some reason the game runs almost entirely smooth at like 120-140 fps. Sometimes gives me 144, sometimes have to set it to 120 but its pretty normal then (the input lag still there, but not 4 second 100 FPS drop.)

I have a 3080, I7 12 core, BenQ 240HZ 1440P and play on 2560x1440. 1GB fiber on Ethernet. DP cable.

**Comment by jeffhill** [comment]
@tonymonsalo The timing in your last video looks correct to me. The delay before the cursor changes is networking delays. The delay *after* the cursor changes is entirely client animation / interpolation / rendering, to make Dota appear to animate smoothly. Could you try the same test in hero demo mode? I'd expect overall latency to be better - but the delay from the cursor changing to the hero appearing to move should be the same.

**Comment by tonymonsalo** [comment]
> @tonymonsalo The timing in your last video looks correct to me. The delay before the cursor changes is networking delays. The delay _after_ the cursor changes is entirely client animation / interpolation / rendering, to make Dota appear to animate smoothly. Could you try the same test in hero demo mode? I'd expect overall latency to be better - but the delay from the cursor changing to the hero appearing to move should be the same.

https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/b6058232-7105-4a8e-8922-8d38e50bff04
https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/a95caaf7-d9c0-49b4-9512-21a5ebbd9e34

@jeffhill Yesterday i found my old SSD with installed dota of 2022 year (7.31 release).
I'm playing counter strike with my friend by hosting "map de_inferno" in my console, and then he connects with "connect <ip>:<port>" in his console. Today i will try to do the same in dota, and will try to measure whatever i can (if i will be able to host/connect)

**Comment by IgorHoruzha** [comment]
@jeffhill @tonymonsalo 
I made some measurements in the lobby and found that it takes approximately ~0.48-0.58 seconds from clicking to the first movement. Is it possible that I made an error in my measurements? Can the movement be measured using the console or another precise measurement method to eliminate human error?

https://github.com/ValveSoftware/Dota2-Gameplay/assets/29505510/8fa85773-be38-4ebb-bdcc-9d80f6677c7e

**Comment by IgorHoruzha** [comment]
> @jeffhill @tonymonsalo I made some measurements in the lobby and found that it takes approximately ~0.48-0.58 seconds from clicking to the first movement. Is it possible that I made an error in my measurements? Can the movement be measured using the console or another precise measurement method to eliminate human factor?
> 
>  first_move_Lycan.mp4 
> UPD. I feel like have also experienced this issue. I thought that my PC was not efficient enough or there was an issue with my Ethernet connection. I was updating it, but it doesn't fully solve the issue, even after about two-three PC changes.

And feel like I still experience it sometimes.

**Comment by tonymonsalo** [comment]
> @jeffhill @tonymonsalo I made some measurements in the lobby and found that it takes approximately ~0.48-0.58 seconds from clicking to the first movement. Is it possible that I made an error in my measurements? Can the movement be measured using the console or another precise measurement method to eliminate human factor?
> 
>  first_move_Lycan.mp4 
> UPD. I feel like have also experienced this issue. I thought that my PC was not efficient enough or there was an issue with my Ethernet connection. I was updating it, but it doesn't fully solve the issue, even after about two-three PC changes.

I don't know, maybe you changed "cl_interp" value in console; or maybe because you use windowed mode 

> Can the movement be measured using the console or another precise measurement method to eliminate human factor?

I was also hoping that there are some better way, but i didn't found it

**Comment by IgorHoruzha** [comment]
@jeffhill @tonymonsalo
[I made some more precise measurements](https://youtu.be/tHLbGZNa-ZQ). 
I used:
> -console -novid +dota_debug_change_cursor_during_pending_order 1 +cl_interp_raito_0

In my test, it takes approximately 0.68 seconds (2.62s - 1.94s) for the hero to start acting. Additionally, the time for changing the cursor back and beginning the action is "the same".

There was almost no difference between 2.62 and 3.25 seconds (0.63s) - just very short movement. However, since 1.94s it took 0.65 seconds (3.27s - 2.62s) to observe the next significant hero movement. This essential movement occurred between 3.25 and 3.27 seconds, lasting only **0.02 seconds**, a very short time. The same behavior was observed for the next hero action.

**Comment by jeffhill** [comment]
The really useful test is to compare in a lobby vs. hero demo or a locally hosted lobby. I would expect that the time it takes for the cursor to change is different - on a server, you'll have some latency, and on a locally hosted lobby the cursor should immediately change. However, the delay from the cursor changing back to the hero moving should be exactly the same between server lobby and hero demo mode / locally hosted lobby.

**Comment by IgorHoruzha** [comment]
@jeffhill @tonymonsalo 

I [have tested](https://www.youtube.com/watch?v=m7jMrNNK-Ns) the typical movements of Lycan in the hero demo mode.

> In My opinion, the minimum error in measurements is **0.03 seconds**. This occurs because Clipchamp sometimes alters the time I enter, [reducing it by one second](https://youtube.com/clip/Ugkx0RwgRSPV_RWWceCFiSf6Gu-Yqd_jQtbM?si=UPzJoLCQvdjyqr49).

The approximate delay per frame change ranges from approximately **0.33 to 0.35** seconds.

Frame changes time:
> 0.90, 1.23, 1.58, 1.93, 2.23, 2.58, 2.91, 3.23, 3.58, 3.91, 4.23, 4.57, 4.90, 5.23, 5.57, 5.90, 6.23

It appears that My ping can indeed influence these numbers. As a result, the delay is about half of what it was during My previous test in the lobby. My usual ping to East/West Europe servers is 23-40ms.

> > @tonymonsalo The timing in your last video looks correct to me. The delay before the cursor changes is networking delays. The delay _after_ the cursor changes is entirely client animation / interpolation / rendering, to make Dota appear to animate smoothly. Could you try the same test in hero demo mode? I'd expect overall latency to be better - but the delay from the cursor changing to the hero appearing to move should be the same.
> 
> https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/b6058232-7105-4a8e-8922-8d38e50bff04 https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/a95caaf7-d9c0-49b4-9512-21a5ebbd9e34
> 
> @jeffhill Yesterday i found my old SSD with installed dota of 2022 year (7.31 release). I'm playing counter strike with my friend by hosting "map de_inferno" in my console, and then he connects with "connect <ip>:<port>" in his console. Today i will try to do the same in dota (with old and maybe new version), and will try to measure whatever i can (if i will be able to host/connect)

@tonymonsalo 
Do You have any demo hero/local measurements? It appears Your tests were conducted in a **server-hosted lobby**, as indicated by the cursor change.

**Comment by tonymonsalo** [comment]
> However, the delay from the cursor changing back to the hero moving should be exactly the same between server lobby

In **demo** mode it is 50ms-90ms to restore cursor, and **50ms** to move after cursor restored. (first two videos - are differently started demo modes)
In a **locally hosted lobby**: cursor restoring takes 80-100ms (maybe it can be 50-100; but i didn't catch such), then, hero starts moving after around **50ms**. 
[venge1_**demo**_50_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/b3e8f0a1-765b-4743-9344-df26073aab13)
[venge2_**demo**_67_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/c7bd7aa1-0978-4451-9e28-31d65f17f442)
[db1_**lobby**_80_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/9e07c4ef-68f2-41f5-990c-010a0b546835)
[db_**lobby**_100_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/6984859e-e543-4569-91d1-bef78544ed3e)

Then i played **turbo game**:
First video (game just started) is showing that hero makes move after **80ms** from cursor restored:
[venge1_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/013298cf-bd60-4950-990b-41aad546ec21)
These videos (almost end of game) shows that this delay is **50ms**:
[venge2_**turbo_5**0](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/1ed4364b-742f-464c-84ca-f8b555bbda03)
[venge3_**turbo_50**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/32a20f8e-b9ac-435a-a805-933a271b1738)

Another **turbo game** shows **80ms** for move after cursor restored:
[db1_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/6503c79b-19af-470e-831f-693d8526261f)
[db2_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/7c500c28-6fbb-48df-b661-4d3d811da2d5)

Looks like sometimes there are happening extra 30ms for moving. 
((and what if these 30ms can also happen during pending_order?))

Hope this revealed something 🙃

**Comment by tonymonsalo** [comment]
> @jeffhill @tonymonsalo
> 
> I [have tested](https://www.youtube.com/watch?v=m7jMrNNK-Ns) the typical movements of Lycan in the hero demo mode.
> 
> > In My opinion, the minimum error in measurements is **~0.03 seconds**. This occurs because Clipchamp sometimes alters the time I enter, [reducing it by one second](https://youtube.com/clip/Ugkx0RwgRSPV_RWWceCFiSf6Gu-Yqd_jQtbM?si=UPzJoLCQvdjyqr49).
> 
> The approximate delay per frame change ranges from approximately **0.33 to 0.35** seconds.
> 
> Frame changes time:
> 
> > 0.90, 1.23, 1.58, 1.93, 2.23, 2.58, 2.91, 3.23, 3.58, 3.91, 4.23, 4.57, 4.90, 5.23, 5.57, 5.90, 6.23
> 
> It appears that My ping can indeed influence these numbers. As a result, the delay is about half of what it was during My previous test in the lobby. My usual ping to East/West Europe servers is 23-40ms.
> 
> > > @tonymonsalo The timing in your last video looks correct to me. The delay before the cursor changes is networking delays. The delay _after_ the cursor changes is entirely client animation / interpolation / rendering, to make Dota appear to animate smoothly. Could you try the same test in hero demo mode? I'd expect overall latency to be better - but the delay from the cursor changing to the hero appearing to move should be the same.
> > 
> > 
> > https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/b6058232-7105-4a8e-8922-8d38e50bff04 https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/a95caaf7-d9c0-49b4-9512-21a5ebbd9e34
> > @jeffhill Yesterday i found my old SSD with installed dota of 2022 year (7.31 release). I'm playing counter strike with my friend by hosting "map de_inferno" in my console, and then he connects with "connect :" in his console. Today i will try to do the same in dota (with old and maybe new version), and will try to measure whatever i can (if i will be able to host/connect)
> 
> @tonymonsalo Do You have any demo hero or local-hosted measurements? It appears Your tests were conducted in a **server-hosted lobby**, as indicated by the cursor change.

Yes, just commented

**Comment by IgorHoruzha** [comment]
> > @tonymonsalo Do You have any demo hero or local-hosted measurements? It appears Your tests were conducted in a **server-hosted lobby**, as indicated by the cursor change.
> 
> People got this issue exactly after 7.3x patch released. And i didn't noticed internet connection issues for the last months (ping is primarily 30-40ms) By the way, what is your PC configuration? Maybe there are some extra-unique hardware incompatibility with whatever in new updates happened

```
Processor 13th Gen Intel(R) Core(TM) i5-13600KF, 3500 MHz, 14 Core(s), 14 Logical Processor(s)
System type 64-bit operating system, x64-based processor
Edition Windows 11 Pro
OS build 22621.3007
Experience Windows Feature Experience Pack 1000.22681.1000.0
BIOS Version/Date American Megatrends Inc. 1604, 12/15/2023
SMBIOS Version 3.5
Motherboard Asus PRIME Z790-P WIFI
Video Card NVIDIA GeForce GT 1030
Ram Kingston FURY DDR5-5200 32768MB PC5-41600 (Kit of 2x16384) Beast AM5 Black (KF552C36BBEK2-32)

Hyper-V - VM Monitor Mode Extensions Yes
Hyper-V - Second Level Address Translation Extensions Yes
Hyper-V - Virtualization Enabled in Firmware No
Hyper-V - Data Execution Protection Yes

Mouse Xtrfy M42 RGB USB White
Keyboard Xtrfy K4 TKL RGB Kailh Red USB White
Monitor 27 DELL UP2716DA (210-AXWI)
```

BIOS settings are modified for now.

**Comment by IgorHoruzha** [comment]
> > @tonymonsalo Do You have any demo hero or local-hosted measurements? It appears Your tests were conducted in a **server-hosted lobby**, as indicated by the cursor change.
> 
> People got this issue exactly after 7.3x patch released. And i didn't noticed internet connection issues for the last months (ping is primarily 30-40ms) By the way, what is your PC configuration? Maybe there are some extra-unique hardware incompatibility with whatever in new updates happened

Possibly, You are right; I'm not ready to claim that I experienced the exact same issue before 7.3x. I just feel like I encountered something that, in my opinion, appeared quite similar.

**Comment by IgorHoruzha** [comment]
> > However, the delay from the cursor changing back to the hero moving should be exactly the same between server lobby
> 
> In **demo** mode it is 50ms-90ms to restore cursor, and **50ms** to hero move after cursor restored. (first two videos - are differently started demo modes) In a **locally hosted lobby**: cursor restoring takes 80-100ms (maybe it can be 50-100; but i didn't catch such), then, hero starts moving after around **50ms**. [venge1_**demo**_50_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/b3e8f0a1-765b-4743-9344-df26073aab13) [venge2_**demo**_67_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/c7bd7aa1-0978-4451-9e28-31d65f17f442) [db1_**lobby**_80_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/9e07c4ef-68f2-41f5-990c-010a0b546835) [db_**lobby**_100_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/6984859e-e543-4569-91d1-bef78544ed3e)
> 
> Then i played **turbo game**: First video (game just started) is showing that hero makes move after **80ms** from cursor restored: [venge1_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/013298cf-bd60-4950-990b-41aad546ec21) These videos (almost end of game) shows that this delay is **50ms**: [venge2_**turbo_5**0](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/1ed4364b-742f-464c-84ca-f8b555bbda03) [venge3_**turbo_50**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/32a20f8e-b9ac-435a-a805-933a271b1738)
> 
> Another **turbo game** shows **80ms** for move after cursor restored: [db1_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/6503c79b-19af-470e-831f-693d8526261f) [db2_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/7c500c28-6fbb-48df-b661-4d3d811da2d5)
> 
> Looks like sometimes there are happening extra 30ms for moving. ((and what if these 30ms can also happen during pending_order?))
> 
> Hope this revealed something 🙃

I apologize for the clarifying question, but did You manually set the localhost server for the lobby? To be honest, I was surprised that lobbies are not hosted locally by default.
![local_host_lobby](https://github.com/ValveSoftware/Dota2-Gameplay/assets/29505510/be09823a-d11c-4093-ab85-7d4a4de5447c)

By the way, what is Your PC configuration?

**Comment by IgorHoruzha** [comment]
> @jeffhill @tonymonsalo
> 
> I [have tested](https://www.youtube.com/watch?v=m7jMrNNK-Ns) the typical movements of Lycan in the hero demo mode.
> 
> > In My opinion, the minimum error in measurements is **~0.03 seconds**. This occurs because Clipchamp sometimes alters the time I enter, [reducing it by one second](https://youtube.com/clip/Ugkx0RwgRSPV_RWWceCFiSf6Gu-Yqd_jQtbM?si=UPzJoLCQvdjyqr49).
> 
> The approximate delay per frame change ranges from approximately **0.33 to 0.35** seconds.
> 
> Frame changes time:
> 
> > 0.90, 1.23, 1.58, 1.93, 2.23, 2.58, 2.91, 3.23, 3.58, 3.91, 4.23, 4.57, 4.90, 5.23, 5.57, 5.90, 6.23
> 
> It appears that My ping can indeed influence these numbers. As a result, the delay is about half of what it was during My previous test in the lobby. My usual ping to East/West Europe servers is 23-40ms.
> 
> > > @tonymonsalo The timing in your last video looks correct to me. The delay before the cursor changes is networking delays. The delay _after_ the cursor changes is entirely client animation / interpolation / rendering, to make Dota appear to animate smoothly. Could you try the same test in hero demo mode? I'd expect overall latency to be better - but the delay from the cursor changing to the hero appearing to move should be the same.
> > 
> > 
> > https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/b6058232-7105-4a8e-8922-8d38e50bff04 https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/a95caaf7-d9c0-49b4-9512-21a5ebbd9e34
> > @jeffhill Yesterday i found my old SSD with installed dota of 2022 year (7.31 release). I'm playing counter strike with my friend by hosting "map de_inferno" in my console, and then he connects with "connect :" in his console. Today i will try to do the same in dota (with old and maybe new version), and will try to measure whatever i can (if i will be able to host/connect)
> 
> @tonymonsalo Do You have any demo hero or local-hosted measurements? It appears Your tests were conducted in a **server-hosted lobby**, as indicated by the cursor change.

@jeffhill @tonymonsalo

I apologize. [My data timing is not exact](https://www.youtube.com/watch?v=hoSm8BNDRY8). Today **I was trying to trigger @tonymonsalo last measurements and found the mistake on My side**. It makes the video longer than in reality when the speed is 0.10x. I'm going to take some new measurements.
In real-time (1x), it takes 0.02-0.03 seconds; in 0.10x mode, it takes approximately 0.32 seconds.




> > However, the delay from the cursor changing back to the hero moving should be exactly the same between server lobby
> 
> In **demo** mode it is 50ms-90ms to restore cursor, and **50ms** to hero move after cursor restored. (first two videos - are differently started demo modes) In a **locally hosted lobby**: cursor restoring takes 80-100ms (maybe it can be 50-100; but i didn't catch such), then, hero starts moving after around **50ms**. [venge1_**demo**_50_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/b3e8f0a1-765b-4743-9344-df26073aab13) [venge2_**demo**_67_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/c7bd7aa1-0978-4451-9e28-31d65f17f442) [db1_**lobby**_80_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/9e07c4ef-68f2-41f5-990c-010a0b546835) [db_**lobby**_100_50](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/6984859e-e543-4569-91d1-bef78544ed3e)
> 
> Then i played **turbo game**: First video (game just started) is showing that hero makes move after **80ms** from cursor restored: [venge1_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/013298cf-bd60-4950-990b-41aad546ec21) These videos (almost end of game) shows that this delay is **50ms**: [venge2_**turbo_5**0](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/1ed4364b-742f-464c-84ca-f8b555bbda03) [venge3_**turbo_50**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/32a20f8e-b9ac-435a-a805-933a271b1738)
> 
> Another **turbo game** shows **80ms** for move after cursor restored: [db1_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/6503c79b-19af-470e-831f-693d8526261f) [db2_**turbo_80**](https://github.com/ValveSoftware/Dota2-Gameplay/assets/20268195/7c500c28-6fbb-48df-b661-4d3d811da2d5)
> 
> Looks like sometimes there are happening extra 30ms for moving. ((and what if these 30ms can also happen during pending_order?))
> 
> Hope this revealed something 🙃

In my opinion I managed to trigger You data(~50-100ms demo mode), by decreasing the amount of cores.
In cases:
1) [2 real cores 2.2HGz](https://www.youtube.com/watch?v=-h1X-cMUiXE)
> [ 2 real cores 2.2HGz source video](https://www.youtube.com/watch?v=qABQ8EzHZME)
2) [3 real cores 2.2HGz](https://www.youtube.com/watch?v=qDUVDV66KKI)
> [3 real cores 2.2HGz source video](https://www.youtube.com/watch?v=flM4PPQG1KA)
3) [4 real cores 1.9HGz](https://www.youtube.com/watch?v=q1uyAGvfbEU)
> [4 real cores 1.9HGZ source video](https://www.youtube.com/watch?v=iH15x9SPGpQ)

[And even 14 real cores 5HGz source video.](https://youtu.be/9yRheJ3uwHw)

This setting was configured in the BIOS.

 **Important: This video was recorded before I discovered the mistake, so they all contain the same error.** I plan to reprocess them ~tomorrow at normal speed but in shorter intervals to ensure more precise timings.
Nevertheless, the cursor appeared inconsistently throughout. And timings are different depends on PC configuration and tries. I'm not going to say that there is no Input lag.

@tonymonsalo
Is it possible that Your measurements has the same mistake? 
What is Your PC configuration?

**Comment by IgorHoruzha** [comment]
> The approximate delay per frame change ranges from approximately **0.33 to 0.35** seconds.

I've just realized that the 0.33-second per frame change is because the Snipping Tool's maximum recording speed is 30 FPS.

```
1 second is 1000ms.
1000/30 = ~33.333....
```

Therefore, it seems there have been an error in my measurements.

**Comment by Bliddddddd3** [comment]
@jeffhill 
I've been researching input lag in s2 games (cs; dota) for 2 years, I wrote a post on what might have a connection to the lag appearing in Dota2 or what might help in finding the cause

https://github.com/ValveSoftware/Dota2-Gameplay/issues/15930

**Comment by jeffhill** [comment]
As of the client update on Monday, we've found and fixed a source of one tick (30ms) of input latency on dedicated servers that would come and go, based on exactly when in the frame your orders would be received by the server. It should now be the case that the delay from "cursor goes back to normal" -> "my hero is moving" should be the same in both hero demo mode and when playing on dedicated servers.

Thanks for all your help, @tonymonsalo!

**Comment by Aeswaaaaa** [comment]
> As of the client update on Monday, we've found and fixed a source of one tick (30ms) of input latency on dedicated servers that would come and go, based on exactly when in the frame your orders would be received by the server. It should now be the case that the delay from "cursor goes back to normal" -> "my hero is moving" should be the same in both hero demo mode and when playing on dedicated servers.
> 
> Thanks for all your help, @tonymonsalo!



> When I wake up I will test again with different values of cl_clock_recvmargin_enable, if I notice a lags pattern I will try to take a video clip through obs, if you need to measure indicators with certain software or inside the dota itself with commands, I can also do it, just let me know what's required of me.

cl_interp_ratio 
cl_clock_recvmargin_enable 

should I leave it at 0? or can I reset it to the default value?

**Comment by Aeswaaaaa** [comment]
> As of the client update on Monday, we've found and fixed a source of one tick (30ms) of input latency on dedicated servers that would come and go, based on exactly when in the frame your orders would be received by the server. It should now be the case that the delay from "cursor goes back to normal" -> "my hero is moving" should be the same in both hero demo mode and when playing on dedicated servers.
> 
> Thanks for all your help, @tonymonsalo!

@jeffhill 

cl_interp_ratio
cl_clock_recvmargin_enable

should I leave it at 0? or can I reset it to the default value?

**Comment by IgorHoruzha** [comment]
@Aeswaaaaa @tonymonsalo 
Was the Input lag fixed for You?
Have experienced this issue #16166?

**Comment by Aeswaaaaa** [comment]
> @Aeswaaaaa @tonymonsalo Would You like to discuss the Input lag issue more? If so was the Input lag fixed for You? Have experienced this issue #16166?

I don't know if these problems are related to each other, but the input lag is present

**Comment by tonymonsalo** [comment]
> @Aeswaaaaa @tonymonsalo 
> 
> Would You like to discuss the Input lag issue more?
> 
> If so was the Input lag fixed for You?
> 
> Have experienced this issue #16166? 

Sorry, provtikal etot comment

**Comment by WongBater** [comment]
Also have input lag issue since act 3 dropped when alt tabbing back into Dota 2

**Comment by YazanKhwaireh** [comment]
> Also have input lag issue since act 3 dropped when alt tabbing back into Dota 2

I also do have the same issue , when i alt tabbing back into data 2 during the picking phase or buying items or moving my hero it takes a time to issue the order , its start happeing after the update of act 3 , my pc is kinda okey i7 8700k with 2080 and 32 RAm and i still encouter this issue

**Comment by WongBater** [comment]
> > Also have input lag issue since act 3 dropped when alt tabbing back into Dota 2
> 
> I also do have the same issue , when i alt tabbing back into data 2 during the picking phase or buying items or moving my hero it takes a time to issue the order , its start happeing after the update of act 3 , my pc is kinda okey i7 8700k with 2080 and 32 RAm and i still encouter this issue

Hey, try using Window Borderless if you are using Exclusive Full screen. It fixed it for me.

**Comment by JCMC05** [comment]
> > > Also have input lag issue since act 3 dropped when alt tabbing back into Dota 2
> > 
> > 
> > I also do have the same issue , when i alt tabbing back into data 2 during the picking phase or buying items or moving my hero it takes a time to issue the order , its start happeing after the update of act 3 , my pc is kinda okey i7 8700k with 2080 and 32 RAm and i still encouter this issue
> 
> Hey, try using Window Borderless if you are using Exclusive Full screen. It fixed it for me.

i already play on window boderless
it helps for some people but the majority this dosent fix

**Comment by JCMC05** [comment]
> > Also have input lag issue since act 3 dropped when alt tabbing back into Dota 2
> 
> I also do have the same issue , when i alt tabbing back into data 2 during the picking phase or buying items or moving my hero it takes a time to issue the order , its start happeing after the update of act 3 , my pc is kinda okey i7 8700k with 2080 and 32 RAm and i still encouter this issue

https://github.com/ValveSoftware/Dota2-Gameplay/issues/19894#issuecomment-2226487827
we are trying to  bring attention to the problem on other posts

**Comment by VetaLioSTV** [comment]
after ACT 3 its re-appeared again. Please fix!

**Comment by khanshaib11** [comment]
match startup lag and alt _ tab lag and resolution changing lag !!! and there is input lag also tried every command on internet but no stable fps i am getting . please fix it dev

**Comment by IgorHoruzha** [comment]
@khanshaib11  Do You still have the issue?

**Comment by VetaLioSTV** [comment]
@IgorHoruzha seems to be i still have, definitely input lag remains visually (with enabled fps/network setting) and seems to be from start it still has that delay, but later in game became stable.

**Comment by IgorHoruzha** [comment]
> @IgorHoruzha seems to be i still have, definitely input lag remains visually (with enabled fps/network setting) and seems to be from start it still has that delay, but later in game became stable.

Let's discuss it in separate issue. Could You open?

**Comment by Fivedota** [comment]
Time to open this case again!

Input lag is back and it's even more horrible then it was one year ago.

Maybe it's good to use this post for troubleshooting (at least the post from feb-mar 2024 since it was at that time the majority of the playerbase had issues).

@jeffhill please look into this again, many complaints at Reddit and also mentioned by several streamers such as Qojqva, Sing, Ramzes

**Comment by tonymonsalo** [comment]
> i am getting input lag on moving character and also increasing 2 , 3 ping on every movement

Looks like true.

[jeffhill](https://github.com/jeffhill) 

![Image](https://github.com/user-attachments/assets/7004478e-0ea4-4d69-bf56-feb41613412f)

**Comment by zengentlemangr** [comment]
1. the only (default) mode that allows full HZ of the monitor is Exclusive Full Screen, other modes when you try to select a resolution is 60hz seems like hardcoded or some issue with windows 11 (anti-kernel level access) or something wierd! I am on AMD RX 7800 XT video card so maybe its an issue with AMD? NVIDIA GPU users please confirm you have this bug re-occure also?

**Comment by zengentlemangr** [comment]
ok fixed with a workaround for me now this is how I solved it:

1. add the game into the windows 11 "graphics" settings:

<img width="426" height="120" alt="Image" src="https://github.com/user-attachments/assets/95519cd7-4584-4c73-a36c-de82884badbe" />

<img width="320" height="348" alt="Image" src="https://github.com/user-attachments/assets/cd5119e8-625a-475d-a83e-919298929e0a" />

2. press "Add desktop app" and find your dota 2 beta bin folder and dota2 executable in either default = C: / Programs x86 / Steam /steamapps / common / dota2beta / game / bin / win64 / dota2.exe or (go to steam storage and see in which drive is the game is located or in the dota 2 right click > properties > local files > show files location) add there the game
<img width="1580" height="984" alt="Image" src="https://github.com/user-attachments/assets/bc6d8df4-b03b-4ad9-998b-d08df72bc3f8" />
3. remove the Optimisations for windowed games for dota 2 (and I would advice also Auto HDR - I dont use HDR but I saw in troubleshooting before that this must be off also)


<img width="1541" height="460" alt="Image" src="https://github.com/user-attachments/assets/f409b534-48ff-4f40-a48a-8f7abc75c405" />

(note here below I by mistake selected power save mode, a big no no, check check check!!!! High performance!!!)
<img width="1594" height="493" alt="Image" src="https://github.com/user-attachments/assets/4bd2af6b-ed63-45af-a45a-111ee6d13c92" />

4. I have it with -coop_fullscreen -force_allow_coop_fullscreen flags

 but I don't know if its needed 

(this basically reveals a check "the full screen will minimize on focus loss) 

which you need to switch off 

if you want to quickly alt tab in and out without the game minimizing 

or if you have multiple monitors  maybe you want to see the game in main screen while you do something else in second screen 

this settings was in linux by default

 but in linux it works great, the game minimizes and opens directly with no lag, but in windows 11 there might be an issue with DWM hidding showing lag for some milliseconds

<img width="281" height="243" alt="Image" src="https://github.com/user-attachments/assets/23a2dfd6-61f8-42c0-84f1-91b1b1e9125c" />

<img width="587" height="134" alt="Image" src="https://github.com/user-attachments/assets/3b1a1b2c-6a43-45a8-a6e5-407edf5b636f" />

result

<img width="1820" height="660" alt="Image" src="https://github.com/user-attachments/assets/8a103f4e-c5ac-4947-9bb0-faded183297f" />
(this says full screen for desktop)

5. I saw the bug come back (there is where most people loose their sh*t having done the whole guides they find

disable variable refresh optimizations from windows "advances display" settings:

<img width="1598" height="215" alt="Image" src="https://github.com/user-attachments/assets/7d39601e-5d97-433e-8a2a-f0ef88c7ef32" />

<img width="1554" height="240" alt="Image" src="https://github.com/user-attachments/assets/02ff577a-30d8-4d74-b224-dbbb31c9c1ed" />

p.s. this is not the same as the switch for variable refresh rate switch in "graphics" display settings, this should be on (the step 5 is optimizations and selective HZ based on sh*t windows 11 thinks you need)

<img width="1544" height="282" alt="Image" src="https://github.com/user-attachments/assets/d5f6f6b2-2a8e-45cf-a657-6ecca2a47cd3" />

p.s. 2: dota 2 is only made to be played with vsync on

this means 1. no fidelityfx upscaling if you reach your monitor top HZ

2. if I recall correctly variable refresh rate works with vsync (at least on AMD) but correct me if I am wrong but I feel vrr working on my end with vsync

**Comment by zengentlemangr** [comment]
I need to do some more research into what of all these modification is the root, cause the point 5 is what finally fixed it seems like windows 11 think dota 2 is not a game and "has" to allow it 60hz in full windowed mode
